### ISSUE TITLE
QASM -> Yao IR

### ISSUE THREAD

**Issue by Roger-luo** [QuantumBFS/YaoQASM.jl#1]
- [x] QASM AST design
- [x] move QASM parser from RBNF
- [ ] Yao IR codegen

---

- [ ] Test IBM openqasm examples
- [ ] Test IBM Q (manually?)

**Comment by GiggleLiu** [comment]
About Yao IR, we should probably first define basic blocks to remove "redundant" information in representation, then any circuit can be flatten to these basic blocks. e.g. we can first `flatten` a block to basic blocks like
* `RepeatedBlock` -> `KronBlock`
* `KronBlock` -> `PutBlock`

Basic block types should include
* PutBlock
* ChainBlock (a function in QASM)
* Concentrator (a subroutine defined on a subset of a register)
* all primitive gates
* ControlBlock

This will make it much easier for compiling, any comments?

**Comment by Roger-luo** [comment]
yes, that's what I think as well. And I think in most cases we only need `PutBlock` since `Kron` can be composition of `PutBlock`, since QASM is mainly just putting gates at given position, this would be straight forward.

**Comment by thautwarm** [comment]
Done except for `barrier`, `reset` and `measure`. Waiting for Liu to provide correspoinding Yao APIs.
